### ISSUE TITLE
isFullyDecrypted + isFullyEncrypted in favor of isDecrypted

### ISSUE THREAD

**Issue by tomholub** [openpgpjs/openpgpjs#945]
Actually, working on this, I think I may have a solution to the earlier `isDecrypted` dilemma when only some packets are decrypted. I think we should have two methods:

```js
// every dev needs to know one of the two, based on context - better be explicit
prv.isFullyDecrypted();
prv.isFullyEncrypted();
```
In this case, `isDecrypted` would be removed from documentation, but still functional.

In v5 one day, the original `isDecrypted` method could be replaced with exception: `isDecrypted deprecated: Use isFullyDecrypted or isFullyEncrypted`

Would be happy to contribute the PR. I'd be thrilled to have a way to be very explicit about this, for easier security reviews.

_Originally posted by @tomholub in https://github.com/openpgpjs/openpgpjs/issues/943#issuecomment-521004839_

**Comment by tomholub** [comment]
Here is my early implementation:
```js
  const getPrvPacketsToCheckKeyEncryption = (k) => {
    if (!k.isPrivate()) {
      throw new Error("Cannot check encryption status of secret keys in a Public Key");
    }
    const prvPackets = k.getKeys().map(k => k.keyPacket)
        .filter(p => p.tag === openpgp.enums.packet.secretKey || p.tag === openpgp.enums.packet.secretSubkey);
    if (!prvPackets.length) {
      throw new Error("This key has no private packets. Is it a Private Key?");
    }
    return prvPackets;
  };

  openpgp.key.Key.prototype.isFullyDecrypted = function () {
    return getPrvPacketsToCheckKeyEncryption(this).every(p => p.isDecrypted() === true);
  };

  openpgp.key.Key.prototype.isFullyEncrypted = function () {
    return getPrvPacketsToCheckKeyEncryption(this).every(p => p.isDecrypted() === false);
  };
```

**Comment by twiss** [comment]
My main worry is that this might imply a higher level of support (and endorsement) for partially decrypted keys than we actually have, or want. For example, `Key.prototype.decrypt` currently throws if it wasn't able to decrypt all keys (decrypting GNU stripped keys is considered successful). Similarly, `Key.prototype.encrypt` now also throws if some keys weren't decrypted.

Based on this, if you only use those functions (without the `keyId` parameter) I think you can assume that a key is either fully encrypted or fully decrypted (and hence a single boolean function is sufficient). I think this is a nice invariant, that we shouldn't break unless someone has a strong need to use partially decrypted keys.

(Also, if someone does need that, there's other stuff we'd need to change, like preferring decrypted subkeys to encrypted ones when signing. That's doable, but again added complexity that I'm not sure is worth it.)

What do you think?

**Comment by tomholub** [comment]
I actually use these functions to exactly disallow the use of partially encrypted keys that I would otherwise not notice, preventing trouble. 

For example you may have a key that has primary key protected with a different pass phrase than subkey. 

You run `key.decrypt(pp_a)` and it decrypts the primary key, but not subkey. Returns false. Maybe at this point the app code offers user to "import the key without decrypting". Key gets imported but it's half decrypted now.

I'm not even sure, if the app dev were to check with `isDecrypted` if it's going to return true or false, without checking the source. And maybe that's a problem. 

This could be prevented by the app author if he was more careful (and knowledgeable) but I imagine we could do a better job at guiding the dev to go the right thing. 

I agree we should not introduce the complexity of adding better support for partially encrypted keys, at all. 

I'm just trying to explore reasonable avenues to make our particular codebase easier to undergo security reviews. I imagine this is a common problem in our domain. In this context, I feel `isDecrypted` may be too ambiguous, as sometimes it means "can I use the secret parts now" and other times it means "please confirm this is fully encrypted". 

> decrypting GNU stripped keys is considered successful

This is a great example where the `isFully` methods could be awesome. Is otherwise encrypted key that has such packet fully encrypted? Yes. Is otherwise decrypted key that has such packet fully decrypted? Yes! :) 

These is just my point of view. Interested to hear yours.

**Comment by twiss** [comment]
> For example you may have a key that has primary key protected with a different pass phrase than subkey.
> 
> You run key.decrypt(pp_a) and it decrypts the primary key, but not subkey. Returns false.

But my point is that this currently throws, not returns false, I think. So you should never end up in that situation where `isDecrypted` is ambiguous.

> This is a great example where the isFully methods could be awesome. Is otherwise encrypted key that has such packet fully encrypted? Yes. Is otherwise decrypted key that has such packet fully decrypted? Yes! :)

But that's already how `isDecrypted` currently behaves - `key.isFullyEncrypted()` would be the same as `!key.isDecrypted()` and `isFullyDecrypted` as `isDecrypted`, so I'm not sure they'd add much. I appreciate that the `isFully` names might be clearer, but as long as `isDecrypted` does the right thing and does what people expect I think it's ok.

**Comment by tomholub** [comment]
> But my point is that this currently throws, not returns false, I think. So you should never end up in that situation where isDecrypted is ambiguous.

Sort of, see this:
```js
let = { privateKeyArmored } = await openpgp.generateKey({userIds: [{email: "hello@user.com"}], passphrase: "my pp"});
let {keys: [k]} = await openpgp.key.readArmored(privateKeyArmored);
await k.primaryKey.decrypt("my pp")
k.isDecrypted(); // true, but not 100% correct. "somewhat decrypted"
```
Good thing it returns true, I think we may have had this conversation before and settled that that's the safer thing to do :+1: although I would prefer that `isDecrypted` would in this case throw also: `Private key is only partially decrypted`.

Another case, which you were referring to:
```js
let = { privateKeyArmored } = await openpgp.generateKey({userIds: [{email: "hello@user.com"}], passphrase: "my pp"});
let {keys: [k]} = await openpgp.key.readArmored(privateKeyArmored);
await k.primaryKey.decrypt("my pp");
await k.decrypt("my pp"); // throw: Key packet is already decrypted.
k.isDecrypted()
// true
```
In both these cases, the app and it's dev will end up thinking that the key is decrypted, which is actually not the case (there are still encrypted packets in it).

Therefore maybe this would not be a security issue, but more of a general correctness issue. If I call `await k.decrypt("my pp")` and it throws `Key packet is already decrypted.` in my face and then I confirm with `k.isDecrypted()` which says `true`, I will be very surprised when I then proceed to use the key in `msg.decrypt(k)` and it throws `Private key not decrypted` or something similar, because the message is actually encrypted for a subkey (which is still encrypted).

Maybe you see these arguments as too weak to introduce `isFullyDecrypted` / `isFullyEncrypted`. Maybe you're right. I did feel it brought tangible benefits when I started using these two in our codebase, and I figured others may benefit the same way ¯\\\_(ツ)\_/¯

> But that's already how isDecrypted currently behaves - key.isFullyEncrypted() would be the same as !key.isDecrypted() and isFullyDecrypted as isDecrypted, so I'm not sure they'd add much. I appreciate that the isFully names might be clearer, but as long as isDecrypted does the right thing and does what people expect I think it's ok.

I don't think it behaves the same. That's why the word `Fully`. Consider:
```
let = { privateKeyArmored } = await openpgp.generateKey({userIds: [{email: "hello@user.com"}], passphrase: "my pp"});
let {keys: [k]} = await openpgp.key.readArmored(privateKeyArmored);
await k.primaryKey.decrypt("my pp");
k.isDecrypted(); // true
!k.isDecrypted(); // false
k.isFullyDecrypted(); // false !!
k.isFullyEncrypted(); // false
```
It's very nuanced. That's why I feel OpenPGP.js, aiming to be a higher-level OpenPGP library, should strive to keep that nuance away from the app developer by providing more precise methods depending on what the dev is trying to figure out.

**Comment by twiss** [comment]
What I was referring to is that, if you only use the high-level `Key.prototype.decrypt`:

```js
let { privateKeyArmored } = await openpgp.generateKey({
  userIds: [{email: "hello@user.com"}],
  passphrase: "passphrase 1",
  subkeys: [{ passphrase: "passphrase 2" }]
});
let {keys: [k]} = await openpgp.key.readArmored(privateKeyArmored);
await k.decrypt("passphrase 1"); // throw: Incorrect key passphrase
k.isDecrypted() // true
```

It's true that `isDecrypted` returns something ambiguous in this case, which is not ideal*, but you should never get to that line of code since the previous line threw, which can be taken as an indication that the key was left in an indeterminate state.**

If you use the low-level `SecretKey.prototype.decrypt`, that's fine, but then I think it's reasonable that you also have to use the low-level `SecretKey.prototype.isDecrypted`, e.g. something along the lines of
`key.getKeys().every(key => key.isDecrypted())` and `key.getKeys().every(key => !key.isDecrypted())`. Of course, it probably make sense for you to add convenience functions for that, but I'm not sure it makes sense to have them on `Key.prototype` built-in.

> I don't think it behaves the same.

Yeah, I just meant within the context of GPG stripped keys, they'd behave the same for those, right? Unless they're also partially decrypted.

(*) If you want to argue that this is a bug that should be fixed, I could get on board with that - perhaps by adding the functions you propose, or perhaps if we get to reimagine the API for v5, a cleaner way to fix this would be to *return a clone of the key in `Key.prototype.decrypt`* - that way, it would either return a fully decrypted key, or throw an error, nothing in between.

(**) You could abuse this API, combined with a try/catch, to incrementally decrypt a key encrypted with multiple passwords, but that's really not what it's intended for (it expects you to pass an array of passwords instead) and the `Incorrect key passphrase` error would make that really confusing, so that use case would be better served by the low-level API.

**Comment by tomholub** [comment]
> It's true that isDecrypted returns something ambiguous in this case, which is not ideal*, but you should never get to that line of code since the previous line threw, which can be taken as an indication that the key was left in an indeterminate state.**

I'm unsure. To me, a thrown error indicates that the key was left unchanged (which may not be true - it could have already decrypted half of it, if the key was imported).

I the app developer may be catching that error, and then giving user options as to what to do next with the key. For us, knowing (fairly) well how the OpenPGP.js api works, we know that's not a good idea. But I'd argue that this is not obvious to many, if not most, people who first start using this library.

> (*) If you want to argue that this is a bug that should be fixed, I could get on board with that - perhaps by adding the functions you propose, or perhaps if we get to reimagine the API for v5, a cleaner way to fix this would be to return a clone of the key in Key.prototype.decrypt - that way, it would either return a fully decrypted key, or throw an error, nothing in between.

This would be 100% amazing. We would probably have to rename `Key.prototype.decrypt(pp)` to `Key.prototype.decrypted(pp)`, just to make sure when people upgrade the lib it breaks in an obvious way so devs can stop to think about it. Either way, this would be I think a real/solid solution to the problem that I was trying to nibble away at with the proposal above!

**Comment by tomholub** [comment]
I know we may or may not end up using these, but I'm adding my latest implementation for completeness, this one handles (skips) gnu-dummy packets.

```js
(() => {
  const getPrvPackets = (k: OpenPGP.key.Key) => {
    if (!k.isPrivate()) {
      throw new Error("Cannot check encryption status of secret keys in a Public Key");
    }
    const prvPackets = k.getKeys().map(k => k.keyPacket).filter(Pgp.key.isPacketPrivate) as PrvPacket[];
    if (!prvPackets.length) {
      throw new Error("This key has no private packets. Is it a Private Key?");
    }
    // only encrypted keys have s2k (decrypted keys don't needed, already decrypted)
    // if s2k is present and it indicates it's a dummy key, filter it out
    // if s2k is not present, it's a decrypted real key (not dummy)
    const nonDummyPrvPackets = prvPackets.filter(p => !p.s2k || p.s2k.type !== 'gnu-dummy');
    if (!nonDummyPrvPackets.length) {
      throw new Error("This key only has a gnu-dummy private packet, with no actual secret keys.");
    }
    return nonDummyPrvPackets;
  };
  openpgp.key.Key.prototype.isFullyDecrypted = function () {
    return getPrvPackets(this).every(p => p.isDecrypted() === true);
  };
  openpgp.key.Key.prototype.isFullyEncrypted = function () {
    return getPrvPackets(this).every(p => p.isDecrypted() === false);
  };
  openpgp.key.Key.prototype.isPacketDecrypted = function (keyId: OpenPGP.Keyid) {
    if (!this.isPrivate()) {
      throw new Error("Cannot check packet encryption status of secret key in a Public Key");
    }
    if (!keyId) {
      throw new Error("No Keyid provided to isPacketDecrypted");
    }
    const [key] = this.getKeys(keyId);
    if (!key) {
      throw new Error("Keyid not found in Private Key");
    }
    return key.keyPacket.isDecrypted() === true;
  };
})();
```
Example usage in app code:
```js
if (!k.isFullyDecrypted() && !k.isFullyEncrypted()) {
  throw new Error("Partially decrypted keys are not supported. Please import a private key that is fully encrypted with the same pass phrase for all secret packets.");
}

**Comment by twiss** [comment]
Fixed by #1182 / #1047.
